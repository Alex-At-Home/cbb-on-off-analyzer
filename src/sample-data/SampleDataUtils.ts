import _ from "lodash";

export class SampleDataUtils {

  /** For snapshot testing, provides a more useful output than a giant list of stats */
  static summarizeEnrichedApiResponse(exampleRawResponse: Record<string, any>, mutableSample: boolean = false) {
    const expectedFields1 = mutableSample ? null : new Set(_.keys(exampleRawResponse));

    return {
      test: (val: any) => val && (val.off_efg ||
          val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg") ||
          val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")
        ),
        //(pick a field that all stat sets have - plus special case for: diagnostics)
      print: (val: any, serialize: ((val: any) => string), indent: ((val: any) => string)) => {
        const expectedFields = expectedFields1 || new Set(_.keys(exampleRawResponse));
        if (val.off_efg) {
          return "{\n" +
            _.chain(val).toPairs().filter(kv => {
              return !expectedFields.has(kv[0]) //derived field)
                || (!kv[1]?.hasOwnProperty("value") && exampleRawResponse[kv[0]]?.hasOwnProperty("value")) //missing value
                || (kv[1]?.hasOwnProperty("value") && _.keys(kv[1]).length > 1) //(has extra properties)
            }).sortBy(0).map(kv => {
              return indent(`${kv[0]}: ` + serialize(kv[1]))
            }).value().join("\n\n") +
          "}";
        } else if (val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg")) {
          return JSON.stringify(_.pick(val, [
            "oRtg", "adjORtg", "adjORtgPlus", "dRtg", "adjDRtg", "adjDRtgPlus"
          ]), null , 3);
        } else if (val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")) {
          return JSON.stringify(_.pick(val, [
            "deltaOffPpp", "deltaOffEfg", "deltaOffAdjEff", "deltaDefPpp", "deltaDefEfg", "deltaDefAdjEff"
          ]), null , 3);
        } else {
          return "???";
        }
      }
    }
  }


  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateLineup(input: Record<string, any>) {
    const lineupAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.toPairs(lineupAsObj).map(kv => {
      return { key: kv[0], ...(kv[1] as Record<string, any>) };
    })
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateTeam(input: Record<string, any>) {
    const teamAsObj = _.transform(_.toPairs(input), (acc, inElKv) => {
      _.forEach(_.toPairs(inElKv[1] as Record<string, any>), (kv) => {
        _.forEach(_.toPairs(kv[1] as Record<string, any>), (deepKv) => {
          if (inElKv[0] != "main") {
            _.set(acc, `${kv[0]}.${deepKv[0]}${inElKv[0]}.value`, deepKv[1]);
          } else {
            _.set(acc, `${deepKv[0]}.${kv[0]}`, deepKv[1]);
          }
        });
      });
    }, {} as Record<string, any>);
    return teamAsObj;
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplatePlayer(input: Record<string, any>) {
    const flatPlayersAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.transform(["baseline", "on", "off"], (acc, v) => {
      acc[v] = { player: {
        buckets: _.chain(flatPlayersAsObj).toPairs().filter(kv => {
          return kv[0].indexOf(v) == 0;
        }).map(kv => {
          const key = kv[0].substring(v.length + 2); //(+2 for ": ")
          return { ...(kv[1]), key: key }
        }).value()
      } };
    }, {} as Record<string, any>);
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseTeam(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().map((kv: [string, any]) => {
        return { key: kv[0], ...(kv[1]) }
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponsePlayer(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().flatMap((kv: [string, any]) => {
        const players = (kv[1]?.player?.buckets || []) as Array<Record<string, any>>;
        return players.map(p => { return { ...p, key: `${kv[0]}: ${p.key}` } });
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseLineup(input: Array<Record<string, any>>, filter?: Set<String>) {
    const preSorted = _.transform(input, (acc, inEl) => {
      const recordId = inEl?.key || "unknown";
      const injectInfo = (k: string, prefix: string) => {
        const topLevelKey = k.substring(prefix.length);
        const realPrefix = prefix || "main";
        if (prefix) {
          const existingObj = _.get(acc, `${topLevelKey}.${recordId}`) || {};
          existingObj[realPrefix] = inEl[k]?.value;
          const sortedObj = _(existingObj).toPairs().sortBy(0).reverse().fromPairs().value()
          _.set(acc, `${topLevelKey}.${recordId}`, sortedObj);
        } else {
          _.set(acc, `main.${topLevelKey}.${recordId}`, inEl[k]);
        }
      };

      if (!filter || filter.has(recordId)) {
        _.keys(inEl).filter(k => k != "key").forEach(k => {
          const stat = _.find([
            "oppo_total_off_", "oppo_total_def_",
            "oppo_off_", "oppo_def_",

            "team_total_off_", "team_total_def_",
            "team_off_", "team_def_",

            "total_off_", "total_def_",
            "off_", "def_"
          ], prefix => k.indexOf(prefix) == 0);
          if (stat) {
            injectInfo(k, stat);
          } else {
            injectInfo(k, "");
          }
        });
      }
    }, {} as Record<string, any>);

    // Sort the keys into some sort of logical order
    return _.chain(preSorted).toPairs().sortBy([
      (kv: [ string , any ]) => {
        const key = kv[0];
        if (key == "main") {
          return 100;
        } else if ((key.indexOf("assist") >= 0) || (key.indexOf("to") >= 0) || (key.indexOf("ast") >= 0)) {
          return 4;
        } else if ((key.indexOf("ft") == 0) || (key.indexOf("fg") == 0) || (key == "efg")) {
          return 1;
        } else if ((key.indexOf("2p") == 0) || (key.indexOf("3p") == 0)) {
          return 2;
        } else if ((key.indexOf("rb") > 0)) {
          return 3;
        } else if ((key.indexOf("adj") >= 0)) {
          return 5;
        } else { //misc other stuff
          return 6;
        }
      },
      (kv: [ string , any ]) => {
        return kv[0];
      },
    ]).fromPairs().value();
  }
}
