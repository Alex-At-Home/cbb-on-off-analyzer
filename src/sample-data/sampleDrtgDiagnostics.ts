export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
