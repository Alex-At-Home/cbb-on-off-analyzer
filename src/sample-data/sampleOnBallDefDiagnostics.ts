/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOnBallDefDiagnostics =
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
};
