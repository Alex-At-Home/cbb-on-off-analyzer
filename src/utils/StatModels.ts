// React imports:
import React from "react";

import {
  OffLuckAdjustmentDiags,
  DefLuckAdjustmentDiags,
} from "./stats/LuckUtils";
import { ORtgDiagnostics, DRtgDiagnostics } from "./stats/RatingUtils";
import { RedBlackTree } from "@collectable/red-black-tree";

export type DivisionStatistics = {
  /** The number of teams in the tier (includes teams in multiple tiers) */
  tier_sample_size: number;
  /** The number of teams in the tier (only teams in their "natural" tier) */
  dedup_sample_size: number;
  /* Sorted list of samples by field name (includes teams in multiple tiers) */
  tier_samples: Record<string, Array<number>>;

  /** Lets you do faster search of what percentile you are in by first looking up with .toFixed(0), or (100*).toFixed*/
  tier_lut: Record<
    string,
    {
      isPct?: boolean; //(whether you need to *100 before applying .toFixed(0))
      lutMult?: number; //(alternative to isPct, let's you specify *10, *100)
      size: number; //(total number of samples in the LUT)
      min: number; //(don't need max, if value missed LUT and is >max then %ile==100, else 1)
      lut: Record<string, Array<number>>; //([0] of the entry is the offset)
      spaces_between?: RedBlackTree.Instance<number, number>; //(if not in the LUT use an optimized binary chop)
    }
  >;

  compression_factor?: number; //(defaults to 1, else we compress tier_lut[field].lut[lookup] by this number)

  /* Sorted list of samples by field name (only teams in their "natural" tier) - used to build the combined files */
  dedup_samples: Record<string, Array<number>>;
};

/** Represents a player, lineup, or team statistic */
export type Statistic = {
  /** The current most relevant number */
  value?: number;
  /** If a number was overridden by manual or luck adjustments, the original */
  old_value?: number;
  /** Provides some context for value */
  extraInfo?: string | React.ReactNode;
  /** Gives some details about how value was derived from old_value */
  override?: string | React.ReactNode;

  /** The number of samples that generated this statistic, if available (assume not) */
  samples?: number;

  /** Background color override (currently only supported for tables, not shadow background) */
  colorOverride?: number;
};
/** Like statistic, but can take an HTML val in its value */
export type StatisticOrRender = {
  /** The current most relevant number */
  value?: number | React.ReactNode | string;
  /** If a number was overridden by manual or luck adjustments, the original */
  old_value?: number | React.ReactNode | string;
  /** Provides some context for value */
  extraInfo?: string | React.ReactNode;
  /** Gives some details about how value was derived from old_value */
  override?: string | React.ReactNode;

  /** The number of samples that generated this statistic, if available (assume not) */
  samples?: number;
};

/** TODO: consider coming up with a list of all possible fields from ES and using a mapped type? */
export type PureStatSet = Record<string, Statistic>;

/** (eg ErAyala) */
export type PlayerCode = string;

/** (eg "Ayala, Eric") */
export type PlayerId = string;

export type PlayerCodeId = { code: PlayerCode; id: PlayerId };

export type RosterEntry = {
  player_code_id?: PlayerCodeId;
  height_in?: number;
  /** Listed jersey number */
  number?: string;
  /** Added by hand to fix errors */
  alt_number?: string;
  /** From NCAA roster - just G, F, C */
  pos?: string;
  /** We inject this periodically based on the most recent stats */
  role?: string;
  /** Ft-In format */
  height?: string;
  /** Fr/So/Jr/Sr */
  year_class?: string;
  gp?: number;
  /** From NCAA roster - "town, ST" in USA or "city, country" if foreign */
  origin?: string;
  lat?: number;
  lon?: number;
};

/** TODO: this doesn't yet support other */
export type OnOffBaselineEnum = "on" | "off" | "baseline";

export type OnOffBaselineOtherEnum = OnOffBaselineEnum | "other";

/** TODO: this doesn't yet support other */
export type OnOffBaselineGlobalEnum = "on" | "off" | "baseline" | "global";

//////////////////////////////////////

// Invidual

/** Non statistical metadata relating to individuals */
export type IndivMetadata = {
  /** eg 'Ayala, Eric' - Id and Key are considered equivalent here  - this field is required */
  key: PlayerId;

  /** For responses that come from ES, a useful indicator of whether the query matched */
  doc_count?: number;

  /** Comes from the ES response */
  player_array?: any; //(long nested type)

  /** Some display info */
  off_title?: string | React.ReactNode;

  /** eg 'Ayala, Eric' gives ErAyala */
  code?: PlayerCode;

  /** Roster info for the player */
  roster?: RosterEntry;

  /** Gets abused to display roster info (height) in the table */
  def_efg?: React.ReactNode;

  /** Gets abused to display roster info (year/class) in the table */
  def_assist?: React.ReactNode;

  /** Gets abused to display assist% */
  def_2primr?: React.ReactNode;
  /** Gets abused to display assist% */
  def_2pmidr?: React.ReactNode;
  /** Gets abused to display assist% */
  def_3pr?: React.ReactNode;

  /** Gets abused to display positional information in the table */
  def_usage?: React.ReactNode;

  /** These can be elements to show spinner while loading */
  off_adj_rapm?: React.ReactNode;
  def_adj_rapm?: React.ReactNode;
  off_adj_rapm_prod?: React.ReactNode;
  def_adj_rapm_prod?: React.ReactNode;

  // Implementation detail for roster stats table:
  onOffKey?: "On" | "Off" | "Baseline" | "Global";

  // These are used in the leaderboard:
  conf?: string;
  team?: string;
  year?: string;
};

/** Derived stats we add to the individual's stat set */
export type IndivEnrichment = {
  /** Positional info derived from statistics */
  role?: string;

  /** Positional diag info - see also IndivPosInfo below */
  posClass?: string;
  posConfidences?: number[];
  posFreqs?: number[];

  /** Luck diags */
  off_luck?: OffLuckAdjustmentDiags;
  /** Luck diags */
  def_luck?: DefLuckAdjustmentDiags;

  /** Off Rating diags */
  diag_off_rtg?: ORtgDiagnostics;
  /** Def Rating diags */
  diag_def_rtg?: DRtgDiagnostics;
};

export type IndivStatSet = PureStatSet & IndivEnrichment & IndivMetadata;

/** Contains stats defining the position role for a given player */
export type IndivPosInfo = {
  posClass: string;
  posConfidences: number[];
  roster?: RosterEntry;
};

//////////////////////////////////////

// Team

/** Non statistical metadata relating to teams */
export type TeamMetadata = {
  /** For responses that come from ES, a useful indicator of whether the query matched */
  doc_count: number;

  /** Only present for global team info */
  roster?: Record<PlayerCode, RosterEntry>;
};

/** Derived stats we add to the team's stat set */
export type TeamEnrichment = {
  // Title info for tables
  off_title?: string | React.ReactNode;
  def_title?: string | React.ReactNode;

  /** Abused to contain the raw net rating (adjusted is in off_net) */
  def_net?: StatisticOrRender;
};

export type TeamStatSet = PureStatSet & TeamEnrichment & TeamMetadata;

//////////////////////////////////////

// Lineup

export type GameInfoStatSet = Record<string, any>; //TODO: model properly

/** Non statistical metadata relating to individuals */
export type LineupMetadata = {
  /** Required, list of codes, _ separated */
  key: string;

  /** For responses that come from ES, a useful indicator of whether the query matched */
  doc_count?: number;

  /** Comes from the ES response */
  players_array?: any; //(long nested type)

  /** From ES (object) or derived by aggregation (list), TODO encode the type */
  game_info?: GameInfoStatSet | Array<GameInfoStatSet>;

  // These are used in the leaderboard:
  conf?: string;
  team?: string;
  year?: string;
  /** Injected info about players for lineup leaderboards */
  player_info?: Record<PlayerId, IndivPosInfo & IndivStatSet>;
};

/** Derived stats we add to the individual's stat set */
export type LineupEnrichment = {
  /** Luck diags */
  off_luck_diags?: OffLuckAdjustmentDiags;
  /** Luck diags */
  def_luck_diags?: DefLuckAdjustmentDiags;

  /** For aggregating lineups with a filter (RAPM specific logic) - this is the unfiltered version */
  all_lineups?: LineupStatSet; //TODO: horrible recursiveness, should fix

  /** The key field when the lineup is an aggregation (TODO: should just rename to key I think?) */
  posKey?: string;

  /** The codes and ids of players in a lineup */
  codesAndIds?: Array<PlayerCodeId>;

  // Title info for tables
  off_title?: string | React.ReactNode;
  def_title?: string | React.ReactNode;

  /** Abused to contain the raw net rating (adjusted is in off_net) */
  def_net?: StatisticOrRender;

  /** Temp hacky flag to optimize some loops during RAPM */
  rapmRemove?: boolean;

  // see also LineupUtils:OnOffLineupStatSet
};

export type LineupStatSet = PureStatSet & LineupEnrichment & LineupMetadata;

//////////////////////////////////////

export type ShotStats = {
  doc_count?: number;
  shot_chart?: {
    buckets: any[];
  };
};

export type ShotStatsModel = {
  on: {
    off: ShotStats;
    def: ShotStats;
  };
  off: {
    off: ShotStats;
    def: ShotStats;
  };
  other: {
    off: ShotStats;
    def: ShotStats;
  }[];
  baseline: {
    off: ShotStats;
    def: ShotStats;
  };
} & {
  onOffMode?: boolean;
  error_code?: string;
};

//////////////////////////////////////

/** For team leaderboard info */
export type TeamInfo = {
  team_name: string;
  gender: string;
  year: string;
  conf: string;
  adj_off: number;
  adj_def: number;
  adj_off_calc: number; // calculated from raw PPP and SoS
  adj_def_calc: number;
  adj_off_calc_30d: number; // calculated from raw PPP and SoS, last 30d
  adj_def_calc_30d: number;

  opponents: Array<{
    oppo_name: string;
    date_str: string;
    date: number;
    team_scored: number;
    oppo_scored: number;
    off_poss: number;
    def_poss: number;
    avg_lead: number;
    location_type: "Home" | "Away" | "Neutral";
    rank: number;
    adj_off: number;
    adj_def: number;
    wab: number;
    wae: number;
  }>;
};

/** For team stats leaderboard info */
export type TeamStatInfo = {
  team_name: string;
  gender: string;
  year: string;
  conf: string;

  stats: PureStatSet;
};

/** Useful type */
export type RosterStatsByCode = Record<PlayerCode, IndivStatSet>;

//////////////////////////////////////

/* Lineup stints */

export type LineupStintScoreInfo = {
  scored: number;
  allowed: number;
};

export type LineupStintTeamStat = {
  total: number;
  early?: number;
  orb?: number;
  ast?: number;
};

export type LineupStintTeamShot = {
  made?: LineupStintTeamStat;
  attempts?: LineupStintTeamStat;
  ast?: LineupStintTeamStat;
  counts?: LineupStintTeamStat;
};

export type LineupStintTeamStats = {
  num_possessions: number;
  pts: number;
  plus_minus: number;
  player_shot_info: Record<string, number>; //(these are weirdly formatted, use with care)
} & Record<string, LineupStintTeamShot | LineupStintTeamStat>;

export type PlayerCodeIdWithStintStats = {
  stats?: LineupStintTeamStats;
} & PlayerCodeId;

export type LineupStintInfo = {
  players: PlayerCodeIdWithStintStats[];
  lineup_id: string;
  start_min: number;
  end_min: number;
  duration_mins: number;
  duration: number;
  team: {
    team: string;
    year: string;
  };
  opponent: {
    team: string;
    year: string;
  };
  score_info: {
    start_diff: number;
    end_diff: number;
    start: LineupStintScoreInfo;
    end: LineupStintScoreInfo;
  };
  team_stats: LineupStintTeamStats;
  opponent_stats: LineupStintTeamStats;
};

//////////////////////////////////////

/** Useful constants */
export class StatModels {
  static emptyIndiv: () => IndivStatSet = () => {
    return { key: "empty", doc_count: 0 } as IndivStatSet;
  };
  static emptyTeam: () => TeamStatSet = () => {
    return { doc_count: 0 } as TeamStatSet;
  };
  static emptyLineup: () => LineupStatSet = () => {
    return { key: "empty", doc_count: 0 } as LineupStatSet;
  };
}
