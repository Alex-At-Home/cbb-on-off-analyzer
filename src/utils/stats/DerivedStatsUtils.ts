import { PureStatSet } from '../StatModels';

/** Similar to PlayTypeUtils, contains methods to build statistics that are fairly simple derivations of what the ES response returns  */
export class DerivedStatsUtils {

    /** "Scramble" (Post ORB) stats - note only valid for team
     * Fields injected (off/def unless specified): _scramble, _scramble_per_orb, _scramble_ppp, _scramble_delta_ppp
    */
    static readonly injectScrambleStats = (stat: PureStatSet, offDef: "off" | "def", toMutate: PureStatSet) => {
      const totalPoss = stat[`total_${offDef}_poss`]?.value || 1;
      const totalOrbs = stat[`total_${offDef}_orb`]?.value || 1;
      const scramblePct = (stat[`total_${offDef}_scramble_poss`]?.value || 0)/totalPoss;
      const scrambleOrbRatio = (stat[`total_${offDef}_scramble_poss`]?.value || 0)/totalOrbs;
      const totalPpp = (stat[`${offDef}_ppp`]?.value || 0); //TODO: depends on player vs team/lineup
      const scramblePpp = (stat[`${offDef}_scramble_ppp`]?.value || 0) ;
      const scramblePppDelta = scramblePpp - totalPpp;

      if (totalPoss > 0) {
         toMutate[`${offDef}_scramble`] = { value: scramblePct };
         toMutate[`${offDef}_scramble_per_orb`] = { value: scrambleOrbRatio };
         toMutate[`${offDef}_scramble_ppp`] = { value: scramblePpp };
         toMutate[`${offDef}_scramble_delta_ppp`] = { value: scramblePppDelta };
      }
      return toMutate;
   };

    /** Transition stats - note only valid for team
     * Fields injected (off/def unless specified): _trans, _trans_ppp, _trans_delta_ppp
    */
     static readonly injectTransitionStats = (stat: PureStatSet, offDef: "off" | "def", toMutate: PureStatSet) => {
      const totalPoss = stat[`total_${offDef}_poss`]?.value || 1;
      const transPct = (stat[`total_${offDef}_trans_poss`]?.value || 0)/totalPoss;
      const totalPpp = (stat[`${offDef}_ppp`]?.value || 0); 
      const transPpp = (stat[`${offDef}_trans_ppp`]?.value || 0);
      const transPppDelta = transPpp - totalPpp;

      if (totalPoss > 0) {
         toMutate[`${offDef}_trans`] = { value: transPct };
         toMutate[`${offDef}_trans_ppp`] = { value: transPpp };
         toMutate[`${offDef}_trans_delta_ppp`] = { value: transPppDelta };
      }
      return toMutate;
   };

    /** Injects some 4-8 factor stats for transition and scrambles - note only valid for team
     * Fields injected (off/def unless specified): _${playType}_to, _${playType}_ftr, _${playType}_3pr, _${playType}_3p, _${playType}_2p
    */
   static readonly injectStatBreakdowns = (stat: PureStatSet, offDef: "off" | "def", playType: "scramble" | "trans", toMutate: PureStatSet) => {
      const totalPoss = stat[`total_${offDef}_${playType}_poss`]?.value || 0;
      const toPct = (stat[`total_${offDef}_${playType}_to`]?.value || 0)/(totalPoss || 1);
      const fga = stat[`total_${offDef}_${playType}_fga`]?.value || 0;
      const ftr = (stat[`total_${offDef}_${playType}_fta`]?.value || 0)/(fga || 1);
      const threePtA = stat[`total_${offDef}_${playType}_3p_attempts`]?.value || 0;
      const threePtR = threePtA/(fga || 1);
      const threePct = (stat[`total_${offDef}_${playType}_3p_made`]?.value || 0)/(threePtA || 1);
      const twoPct = (stat[`total_${offDef}_${playType}_2p_made`]?.value || 0)/
          (stat[`total_${offDef}_${playType}_2p_attempts`]?.value || 1);

      if (totalPoss > 0) {
         toMutate[`${offDef}_${playType}_to`] = { value: toPct };
         toMutate[`${offDef}_${playType}_ftr`] = { value: ftr };
         toMutate[`${offDef}_${playType}_3pr`] = { value: threePtR };
         toMutate[`${offDef}_${playType}_3p`] = { value: threePct };
         toMutate[`${offDef}_${playType}_2p`] = { value: twoPct };
      }
      return toMutate;
   };

   /** Unadjusted pace of play - valid for player or team
     * Fields injected: tempo
   */
   static readonly injectPaceStats = (stat: PureStatSet, toMutate: PureStatSet, isPlayer: Boolean) => {
      const totalOffPoss = (isPlayer ? stat[`off_team_poss`]?.value : stat[`off_poss`]?.value) || 0;
      const totalDefPoss = (isPlayer ? totalOffPoss : stat[`def_poss`]?.value) || 0;      
      const totalTime = stat[`duration_mins`]?.value || 0;
      const possPer40 = 0.5*(totalOffPoss + totalDefPoss) / (totalTime/40);
      if (totalTime) {
         toMutate[`tempo`] = { value: possPer40 };
      };
      return toMutate;
   }

   /** Injects all the derived stats in one go */
   static readonly injectTeamDerivedStats = (stat: PureStatSet, toMutate: PureStatSet) => {
      DerivedStatsUtils.injectScrambleStats(stat, "off", toMutate);
      DerivedStatsUtils.injectScrambleStats(stat, "def", toMutate);
      DerivedStatsUtils.injectStatBreakdowns(stat, "off", "scramble", toMutate);
      DerivedStatsUtils.injectStatBreakdowns(stat, "def", "scramble", toMutate);
      DerivedStatsUtils.injectTransitionStats(stat, "off", toMutate);
      DerivedStatsUtils.injectTransitionStats(stat, "def", toMutate);
      DerivedStatsUtils.injectStatBreakdowns(stat, "off", "trans", toMutate);
      DerivedStatsUtils.injectStatBreakdowns(stat, "def", "trans", toMutate);
      DerivedStatsUtils.injectPaceStats(stat, toMutate, false);
      return toMutate;
   }

}