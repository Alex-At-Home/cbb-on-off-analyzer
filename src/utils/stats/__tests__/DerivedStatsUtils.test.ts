import { DerivedStatsUtils } from '../DerivedStatsUtils';
import _ from 'lodash';

import { sampleTeamStatsResponse } from "../../../sample-data/sampleTeamStatsResponse"
import { PureStatSet } from '../../StatModels';

describe("DerivedStatsUtils", () => {
   const teamData = _.assign(
      sampleTeamStatsResponse.responses[0].aggregations.tri_filter.buckets as { on: any, off: any, baseline: any },
      { global: {}, onOffMode: true }
    );

   /** Tests all the methods in one go */
   test("DerivedStatsUtils.injectDerivedStats", () => {
      const derivedStats = DerivedStatsUtils.injectDerivedStats(teamData.baseline, {} as PureStatSet);
      expect(derivedStats).toMatchSnapshot();
   });
});
