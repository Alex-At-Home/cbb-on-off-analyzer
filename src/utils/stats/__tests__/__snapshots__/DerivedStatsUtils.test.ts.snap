// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DerivedStatsUtils DerivedStatsUtils.injectDerivedStats 1`] = `
Object {
  "def_scramble": Object {
    "value": 0.08527071563674632,
  },
  "def_scramble_2p": Object {
    "value": 0.4766355140186916,
  },
  "def_scramble_3p": Object {
    "value": 0.42857142857142855,
  },
  "def_scramble_3pr": Object {
    "value": 0.24647887323943662,
  },
  "def_scramble_delta_ppp": Object {
    "value": 8.196096883864556,
  },
  "def_scramble_ftr": Object {
    "value": 0.4507042253521127,
  },
  "def_scramble_per_orb": Object {
    "value": 0.583711708026543,
  },
  "def_scramble_ppp": Object {
    "value": 104.25568362101016,
  },
  "def_scramble_to": Object {
    "value": 0.14088605894731104,
  },
  "def_trans": Object {
    "value": 0.19610764055742433,
  },
  "def_trans_2p": Object {
    "value": 0.5286624203821656,
  },
  "def_trans_3p": Object {
    "value": 0.35294117647058826,
  },
  "def_trans_3pr": Object {
    "value": 0.4311594202898551,
  },
  "def_trans_delta_ppp": Object {
    "value": -5.395533808941735,
  },
  "def_trans_ftr": Object {
    "value": 0.42028985507246375,
  },
  "def_trans_ppp": Object {
    "value": 90.66405292820387,
  },
  "def_trans_to": Object {
    "value": 0.18867924528301885,
  },
  "off_scramble": Object {
    "value": 0.10780838866433026,
  },
  "off_scramble_2p": Object {
    "value": 0.5755395683453237,
  },
  "off_scramble_3p": Object {
    "value": 0.28888888888888886,
  },
  "off_scramble_3pr": Object {
    "value": 0.24456521739130435,
  },
  "off_scramble_delta_ppp": Object {
    "value": 8.046930109156818,
  },
  "off_scramble_ftr": Object {
    "value": 0.47282608695652173,
  },
  "off_scramble_per_orb": Object {
    "value": 0.6415989205037649,
  },
  "off_scramble_ppp": Object {
    "value": 115.22073848662431,
  },
  "off_scramble_to": Object {
    "value": 0.12057984260228126,
  },
  "off_trans": Object {
    "value": 0.24058738565238325,
  },
  "off_trans_2p": Object {
    "value": 0.6779661016949152,
  },
  "off_trans_3p": Object {
    "value": 0.30935251798561153,
  },
  "off_trans_3pr": Object {
    "value": 0.439873417721519,
  },
  "off_trans_delta_ppp": Object {
    "value": 0.0905502377016063,
  },
  "off_trans_ftr": Object {
    "value": 0.6708860759493671,
  },
  "off_trans_ppp": Object {
    "value": 107.2643586151691,
  },
  "off_trans_to": Object {
    "value": 0.16609965979587754,
  },
  "tempo": Object {
    "value": 83.16,
  },
}
`;
